tubular/dates.py

```diff
@@ -49,7 +49,7 @@ def cast_columns_to_datetime(self, X: pd.DataFrame) -> pd.DataFrame:
 
         self.check_columns_are_date_or_datetime(X)
 
-        temp = X[self.columns].copy()
+        temp = X[self.columns]
```

Comment:
Q: Looking at these sorts of lines:

> Therefore, to prevent this I have enforced views to make use of CoW with the following: `df = df[df.columns]`

Are you saying that without these lines (which are copies) then frames will update by reference, even though `pd.options.mode.copy_on_write` is set to `True`? Because if so, copy on write isn't doing what it says. `X[self.columns]` is making a copy I think (would need to profile), but that defeats the whole purpose of the refactor!

Reply:
I had a play and just changing:

```python
X_y = X[X.columns]

X_y["_temporary_response"] = y.to_numpy()
```

to

```python
X_y = X.assign(_temporary_response = y)
```

in `_combine_X_y` alongside removing `X_temp = X[X.columns]` in `BaseTransformer` decreased runtime by 25%:

```bash
# Unchanged
Tubular mean CPU time and SD
0.024680734599999973
0.005784166446244637
***Use cprofile to inspect transform calls
Overall Runtime:  0.040817475857000005

# Above changed
***Use time function to inspect transform calls
Tubular mean CPU time and SD
0.018341116700000028
0.006091312706994393
***Use cprofile to inspect transform calls
Overall Runtime:  0.029613771222
```

which I make to be (0.024 - 0.018)/0.024 = 25%, and that's before profiling memory (I will add these results). All of the relevant tests pass and, crucially, additional columns aren't added to the input dataframes. There are some tests that fail, but these are tests that inspect whether copies were made, and the whole point is that we are placing all of this onto pandas copy-on-write methodology.

Most important of all though is that if the plan is migrate to the `DataFrameAPI` (see #185) then we need to have as clear a layout of code as possible, with just the operations performed - no directives to make defensive copies. In the above case, we could just reparameterise the function to use [`DataFrame.assign`](https://data-apis.org/dataframe-api/draft/API_specification/dataframe_object.html#dataframe_api.DataFrame.assign) and the code should (🤞) just work with polars.

Reply:
OK, some progress.

In the pandas docs I came across the sections [Returning a view versus a copy](https://pandas.pydata.org/pandas-docs/version/0.13/indexing.html#returning-a-view-versus-a-copy), which states:

> The rules about when a view on the data is returned are entirely dependent on NumPy. Whenever an array of labels or a boolean vector are involved in the indexing operation, the result will be a copy. With single label / scalar indexing and slicing, e.g. `df.ix[3:6]` or `df.ix[:, 'A']`, a view will be returned.

(docs seem to be a little out of date as `.ix` [has been replaced by](https://pandas.pydata.org/pandas-docs/version/0.23/generated/pandas.DataFrame.ix.html) `.iloc`).

So the fact that `X[self.columns]` returns a copy is expected behaviour given the rules of numpy. For this PR, this means that we should prefer:

```python
def transform(self, X: pd.DataFrame) -> pd.DataFrame:
       
        ...

        # Create a view of X, permitting copy-on-write semantics.  Via scalar indexing a view will be returned.
        X_view = X.iloc[:]

        return X_view
```

Over

```python
def transform(self, X: pd.DataFrame) -> pd.DataFrame:
       
        ...

        # Create a view of X, permitting copy-on-write semantics.  Via scalar indexing a view will be returned.
        X_copy = X[self.columns]

        return X_copy
```

A brief check on the `capper` example above shows that the original dataframe is not modified, but running the `profile_tubular.py` script demonstrates the ~25% time saving (and that's before we look at memory).

For this ticket then, shall we just add a test to check that the original frame is not updated upon applying a transform, then employ the `X.iloc` style to prevent triggering the copy?

However, this indexing resulting in copies raises more questions than it closes, because if an array-backed indexing operation results in a copy, then all of the lines below are copying the dataframe (!). I will guarantee that this is why the timing profiling is dominated by indexing operations (`__getitem__`, `_take_with_is_copy`, `take`). Tackling this is beyond the scope of this ticket, but it is worth thinking about decreasing this runtime burden. In order to do so with confidence, we need some checks that input frames aren't modified.

```bash
(py38) @adamsardar ➜ /workspaces/tubular/tubular (181-migrate-pandas-usage-to-copy-on-write-and-remove-copy-arg-from-transforms) $ grep -F '[self.columns' ./*
./base.py:        X[self.new_column_names] = getattr(X[self.columns], self.pd_method_name)(
./capping.py:        numeric_column_types = X[self.columns].apply(
./comparison.py:        X[self.new_col_name] = X[self.columns[0]] == X[self.columns[1]]
./dates.py:        temp = X[self.columns]
./dates.py:        temp = X[self.columns]
./dates.py:        if (pd.isna(row[self.columns[0]])) or (pd.isna(row[self.columns[1]])):
./dates.py:        age = row[self.columns[1]].year - row[self.columns[0]].year
./dates.py:            if (row[self.columns[1]].month, row[self.columns[1]].day) < (
./dates.py:                row[self.columns[0]].month,
./dates.py:                row[self.columns[0]].day,
./dates.py:        elif age < 0 and (row[self.columns[1]].month, row[self.columns[1]].day) > (
./dates.py:            row[self.columns[0]].month,
./dates.py:            row[self.columns[0]].day,
./dates.py:            temp[self.columns[1]] - temp[self.columns[0]]
./dates.py:            X[self.columns[0]],
./dates.py:                temp[self.columns[0]].dt,
./dates.py:                temp[self.columns[0]].dt,
./dates.py:        if not (X[self.columns[0]] <= X[self.columns[2]]).all():
./dates.py:            lower_comparison = X[self.columns[0]] <= X[self.columns[1]]
./dates.py:            lower_comparison = X[self.columns[0]] < X[self.columns[1]]
./dates.py:            upper_comparison = X[self.columns[1]] <= X[self.columns[2]]
./dates.py:            upper_comparison = X[self.columns[1]] < X[self.columns[2]]
./misc.py:        X[self.columns] = self.value
./misc.py:        X[self.columns] = X[self.columns].astype(self.dtype)
./nominal.py:        OneHotEncoder.fit(self, X=X[self.columns], y=y)
./nominal.py:        X_transformed = OneHotEncoder.transform(self, X[self.columns])
./numeric.py:        numeric_column_types = X[self.columns].apply(
./numeric.py:            if (X[self.columns] <= -1).sum().sum() > 0:
./numeric.py:                X[new_column_names] = np.log1p(X[self.columns])
./numeric.py:                X[new_column_names] = np.log1p(X[self.columns]) / np.log(self.base)
./numeric.py:            if (X[self.columns] <= 0).sum().sum() > 0:
./numeric.py:                X[new_column_names] = np.log(X[self.columns])
./numeric.py:                X[new_column_names] = np.log(X[self.columns]) / np.log(self.base)
./numeric.py:        if not pd.api.types.is_numeric_dtype(X[self.columns[0]]):
./numeric.py:            msg = f"{self.classname()}: {self.columns[0]} should be a numeric dtype but got {X[self.columns[0]].dtype}"
./numeric.py:            X[self.columns[0]].to_numpy(),
./numeric.py:        is_numeric = X[self.columns].apply(pd.api.types.is_numeric_dtype, axis=0)
./numeric.py:        numeric_column_types = X[self.columns].apply(
./numeric.py:        self.scaler.fit(X[self.columns])
./numeric.py:        X[self.columns] = self.scaler.transform(X[self.columns])
./numeric.py:            [self.columns[col_idx] for col_idx in interaction_combination]
./numeric.py:        numeric_column_types = X[self.columns].apply(
./numeric.py:            if 0 < self.n_components <= min(X[self.columns].shape):
./numeric.py:                msg = f"{self.classname()}: n_components {self.n_components} must be between 1 and min(n_samples {X[self.columns].shape[0]}, n_features {X[self.columns].shape[1]}) is {min(X[self.columns].shape)} with svd_solver '{self.svd_solver}'"
./numeric.py:        self.pca.fit(X[self.columns])
./numeric.py:        X[self.feature_names_out] = self.pca.transform(X[self.columns])
./strings.py:        X[self.new_column_name] = getattr(X[self.columns[0]].str, self.pd_method_name)(
./strings.py:            X[self.columns].astype(str).apply(self.separator.join, axis=1)
```

Reply:
Updated to use view [here](https://github.com/lvgig/tubular/pull/197/commits/33f347f373acfcd71f99291b9d1b46e7e5959a55)
Add test to check original df not changed [here](https://github.com/lvgig/tubular/pull/197/commits/32d917b6e181136262b1607c3702e16a27d51a9b)